### PR TITLE
PEMeta Service: PEV based PE extractor

### PR DESCRIPTION
**Pull request by boddumanohar**

### CODE REVIEW

**Comment by boddumanohar** [comment]
solves #152

File: src/main/scala/org/holmesprocessing/totem/driver/driver.scala [hunk 1]
```diff
@@ -5,19 +5,20 @@ import java.util.concurrent.{Executors, ExecutorService}
 import akka.actor.{ActorRef, ActorSystem, Props}
 import org.holmesprocessing.totem.actors._
 import org.holmesprocessing.totem.services.asnmeta.{ASNMetaSuccess, ASNMetaWork}
+import org.holmesprocessing.totem.services.cfg.{cfgSuccess, cfgWork}
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/driver/driver.scala:8]: really appreciate you reordering all these. Thanks!

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+RUN apk add --no-cache \
+        git \
+        && go get github.com/julienschmidt/httprouter \
+        && git clone -b testing https://github.com/boddumanohar/libpe.git \
+        && rm -rf /var/cache/apk/*
+
+RUN apk add --no-cache \
+        bash \
+        build-base \
+            openssl-dev \
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile]: indentation issue

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	    make \
+	    g++ \
+        && rm -rf /var/cache/apk/*
+
+# Clean Up
+RUN apk del --purge \
+        build-base \
+        && rm -rf /var/cache/apk/* yara-3.5.0
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile]: you are removing yara-3.5.0 but have not installed it. Looks like a copy over from a previous file you were working on.

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,50 @@
+FROM golang:alpine
+
+# Create folder
+RUN mkdir -p /service
+WORKDIR /service
+
+# Get Golang and PEV dependencies
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile]: please follow the format for the other docker files. Following the same layers makes the build process faster due to Docker caching. 

Take a look at virustotal's service for an example.

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/README.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
+# PEMETA service for Holmes-Totem
+
+## Description
+
+This service extracts meta information about a Windows Portable Executable file.
+
+### Output
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/README.md]: great job on the output! thanks!

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+RUN apk add --no-cache \
+        git \
+        && go get github.com/julienschmidt/httprouter \
+        && git clone -b testing https://github.com/boddumanohar/libpe.git \
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile]: did you make a pull request to the PEV guys?

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/pev.go [hunk 7]
```diff
@@ -0,0 +1,575 @@
+package main
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/pev.go]: if this is your code, I would recommend naming it PEMeta.go or something similar.

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/pev.go [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+
+type Config struct {
+	HTTPBinding string
+}
+
+var (
+	config   *Config
+	info     *log.Logger
+	metadata Metadata = Metadata{
+		Name:        "PEinfo",
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/pev.go]: meta data needs to be updated. PEinfo should be PEMeta

**Review by webstergd — CHANGES_REQUESTED** [review]
This looks like some solid work. Great job! I have some small requests that I would like cleaned up before I begin testing. All in all, this looks like it will be a great addition.

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile (hunk 6 above)
Comment (boddumanohar) [reply src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile]: Yes. They reviewed yesterday and asked for changes in design (i.e move each functionality to seperate files). Today, for some time I will work on that. If that is merged, then I will make an update pull request with orginal pev as dependency.

Comment (webstergd) [reply src/main/scala/org/holmesprocessing/totem/services/pemeta/Dockerfile]: That is really great to hear that you are contributing to the other project as well. Good job and good luck with the pull request to pev!

**Comment by webstergd** [comment]
@boddumanohar is making a pull request to the PEV guys. Will see what they say and them proceed with the pull based on how pev wants to handle the additions. Hopefully pev wants to join in the fun

**Comment by boddumanohar** [comment]
with the last 2 commits, PEMETA now supports "__resources__" and "__imphash__"

I've been working with the PEV guys and they just confirmed that interfaces are okay. And I started fixing bugs on memory leaks and other optimization stuff. So from now on, mostly no changes will be made to names that we are using in our PEMETA.

**Comment by webstergd** [comment]
I just checked your work at libpe. Awesome job. Please let me know when that is done and we will merge this in. This should be a major improvement.

**Comment by webstergd** [comment]
close #152

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/README.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
+# PEMETA service for Holmes-Totem
+
+## Description
+
+This service extracts meta information about a Windows Portable Executable file. The analyser library used in PEV's [libpe](https://github.com/merces/libpe). PEV is PE file analysis toolkit. Its feature rich, fast and is written C Programing langauge. This pemeta services wraps all the features of PEV's libpe using golang's cgo.
+
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/README.md:6]: would be good to add your table showing the speed PEMeta runs versus peinfo and peinfov2. As you added concurrency, probably need to rerun your speed tests.

File: src/main/scala/org/holmesprocessing/totem/services/pemeta/service.conf.example [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+	"HTTPBinding": ":8080"
```
Comment (webstergd) [inline src/main/scala/org/holmesprocessing/totem/services/pemeta/service.conf.example]: should have service.conf.example match the new format you did for the other services

**Comment by boddumanohar** [comment]
This version of PEMETA uses original LIBPE as dependency. I also updated README and which shows updated performance metrics. 

according to the table, PEMETA is slower than PEINFO v2. This is because `ccsetup514.exe` took more time to extract resources.

PEMETA is ready for testing.

**Review by webstergd — APPROVED** [review]

**Pull request merged**